src/tasks/base.py

```diff
@@ -0,0 +1,184 @@
+from typing import Tuple, List, Optional
+from abc import ABC, abstractmethod
+
+import torch
+from pytorch_lightning import LightningModule
+from omegaconf import DictConfig
+
+from src.constructor.constuctor import Constructor
+
+
+class BaseTask(LightningModule, ABC):
+    """An abstract class that represent main methods of tasks."""
+
+    def __init__(self, hparams: DictConfig):
+        """Init BaseTask.
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__()
+        self.save_hyperparameters(hparams)
+        self.__constructor = Constructor(hparams)
+        self._metric_manager = self.__constructor.configure_metrics_manager()
+        self._criterion = self.__constructor.configure_losses()
+        self._hparams = self.__constructor.hparams
+        self.__input_shapes = self._hparams.input_shapes
+        self.example_input_array = [torch.rand(1, *shape) for _, shape in self.__input_shapes.items()]
+
+    @abstractmethod
+    def forward(self, *args, **kwargs):
+        """Abstract forward method for validation an test."""
+        pass
+
+    @abstractmethod
+    def forward_with_gt(batch: dict) -> dict:
+        """Abstract forward method for training(with ground truth labels)."""
+        pass
+
+    @abstractmethod
+    def training_step(self, *args, **kwargs):
```

Comment:
I think you don't need to add these abstract methods because they should be overrided in child classes directly from `pl.Module`